common/src/ethereum/erc_20.rs

```diff
@@ -0,0 +1,150 @@
+use ethers::prelude::{signer::SignerMiddlewareError, Http, Lazy, Middleware, ProviderError};
+use ethers::types::{Address, U256};
+use ethers::{
+    abi::Abi,
+    contract::{AbiError, Contract, ContractError},
+    providers::Provider,
+};
+
+use crate::ethereum;
+
+const ABI: &str = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/abis/ERC-20.json"));
+
+static WETH_ADDRESS: Lazy<Address> =
+    Lazy::new(|| Address::from_str("0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2").unwrap());
+static DAI_ADDRESS: Lazy<Address> =
+    Lazy::new(|| Address::from_str("0x6b175474e89094c44da98b954eedeac495271d0f").unwrap());
+static USDC_ADDRESS: Lazy<Address> =
+    Lazy::new(|| Address::from_str("0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48").unwrap());
+static USDT_ADDRESS: Lazy<Address> =
+    Lazy::new(|| Address::from_str("0xdac17f958d2ee523a2206206994597c13d831ec7").unwrap());
+
+pub struct ERC20<M> {
+    contract: Contract<M>,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error<M: Middleware> {
+    #[error(transparent)]
+    Provider(#[from] ProviderError),
+    #[error(transparent)]
+    Contract(#[from] ContractError<M>),
+    #[error(transparent)]
+    Abi(#[from] ethers::abi::Error),
+    #[error(transparent)]
+    ContractAbi(#[from] AbiError),
+    #[error(transparent)]
+    SignerMiddleware(#[from] SignerMiddlewareError<Provider<Http>, ethereum::Wallet>),
+}
+
+impl<M> ERC20<M>
+where
+    M: Middleware,
+    Error<M>: From<<M as Middleware>::Error>,
+{
+    pub fn new(client: impl Into<Arc<M>>, address: Address) -> Self {
+        let abi: Abi = serde_json::from_str(ABI).expect("The ABI is valid");
+        let contract = Contract::new(address, abi, client);
+
+        Self { contract }
+    }
+
+    pub async fn get_symbol(&self) -> Result<String, Error<M>> {
+        let symbol = self
+            .contract
+            .method("symbol", ())
+            .map_err(ContractError::from)?
+            .call()
+            .await?;
+
+        Ok(symbol)
+    }
+
+    pub async fn get_decimals(&self) -> Result<u8, Error<M>> {
+        let symbol = self
+            .contract
+            .method("decimals", ())
+            .map_err(ContractError::from)?
+            .call()
+            .await?;
+
+        Ok(symbol)
+    }
+
+    pub async fn get_allowance(&self, owner: Address, spender: Address) -> Result<U256, Error<M>> {
+        let allowance = self
+            .contract
+            .method("allowance", (owner, spender))
+            .map_err(ContractError::from)?
+            .call()
+            .await?;
+
+        Ok(allowance)
+    }
+
+    pub fn approve(&self, spender: Address, amount: U256) -> Result<ContractCall<M, ()>, Error<M>> {
+        self.contract
+            .method("approve", (spender, amount))
+            .map_err(Error::ContractAbi)
+    }
+}
+
+#[derive(Debug)]
+pub enum Token {
+    WETH,
+    DAI,
+    USDC,
+    USDT,
+    Other(Option<Address>), // e.g 0x2b591e99afe9f32eaa6214f7b7629768c40eeb39
```

Comment:
Can we get rid of the `Option` here? Seems like we shouldn't need it.

Reply:
When you choose `other` in the (cli) list as your token, you'd get an `Other(None)`, you can instantiate an `Other(0x00)` but `None` is better.

As we can pass `--token <addr>` to cli argument, we need to support both `Other` with and without a specified address.

Reply:
Yeah I understand that's how it works right now, but I don't think it makes sense. You are using the type to represent a token as well as for parsing a multi-select where one of the items is not a token ("OTHER"). I don't think the `from_str` instance should accept "OTHER" as a valid `Token`. Then you shouldn't have this problem.

Reply:
Ok, removed the empty case altogether https://github.com/radicle-dev/radicle-cli/commit/f4007e459f8ce654f3ca58ce48dc1f9910838534#diff-5b91c481d0947d833cc80f7e2abb8a4f92ee01e5f225e0c526772affd43acc47R108-R113

Reply:
Great!